src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -594,7 +594,15 @@ internal virtual T ReadWithQuotes(ref Utf8JsonReader reader)
         }
 
         internal virtual void WriteWithQuotes(Utf8JsonWriter writer, [DisallowNull] T value, JsonSerializerOptions options, ref WriteStack state)
-            => ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(TypeToConvert, this);
+        {
+            if (typeof(T) == typeof(string))
```

Comment:
Since `string` is a reference type, this check cannot be optimized by the JIT. It's likely then that users registering a custom string comparer might experience unexpected performance regressions when serializing dictionary keys. Would it be possible to run a few benchmarks, and if performance difference is substantial, replace this with a cached flag?

Reply:
Would we need to add similar fallbacks for other primitive types?

Reply:
So my understanding is that a custom converter for this case is only allowed for `string`, which is also true for 5.0? (i.e. Int32 not supported; the custom converter is just not called). A quick appears to verify this.

Reply:
Benchmarks comparing caching or not caching the string-type check seem inconclusive. I'll switch to a cached value to be safe.

<details>
<summary>Benchmark code (click to view)</summary>

```cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;
using BenchmarkDotNet.Attributes;

namespace MiscBenchmarks
{
    [MemoryDiagnoser]
    public class DictionaryStringCustomConverter
    {
        [Params(10, 100, 1000)]
        public int Count;

        private Dictionary<string, string> _stringKeyDict;

        private Dictionary<int, string> _intKeyDict;

        private readonly JsonSerializerOptions _options = new() { Converters = { new CustomStringConverter() } };

        [GlobalSetup]
        public void Setup()
        {
            _stringKeyDict = new Dictionary<string, string>();
            _intKeyDict = new Dictionary<int, string>();

            for (int i = 0; i < Count; i++)
            {
                _stringKeyDict[$"{i}"] = $"{i}";
                _intKeyDict[i] = $"{i}";
            }
        }

        [Benchmark]
        public void SerializeStringKeyDict() => JsonSerializer.Serialize(_stringKeyDict, _options);

        [Benchmark]
        public void SerializeIntKeyDict() => JsonSerializer.Serialize(_intKeyDict, _options);

        internal sealed class CustomStringConverter : JsonConverter<string>
        {
            public override string Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
                => reader.GetString();

            public override void Write(Utf8JsonWriter writer, string value, JsonSerializerOptions options)
                => writer.WriteStringValue(value);
        }
    }
}
```
</details>

NET 5.0

|                 Method | Count |        Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |------ |------------:|----------:|----------:|-------:|------:|------:|----------:|
| SerializeStringKeyDict |    10 |    814.6 ns |   6.82 ns |   5.32 ns | 0.0801 |     - |     - |     336 B |
|    SerializeIntKeyDict |    10 |    718.9 ns |  10.66 ns |   9.45 ns | 0.0801 |     - |     - |     336 B |
| SerializeStringKeyDict |   100 |  6,566.8 ns |  70.65 ns |  62.63 ns | 0.5035 |     - |     - |    2136 B |
|    SerializeIntKeyDict |   100 |  5,717.8 ns |  47.02 ns |  39.26 ns | 0.5035 |     - |     - |    2136 B |
| SerializeStringKeyDict |  1000 | 67,088.4 ns | 527.79 ns | 440.72 ns | 5.6152 |     - |     - |   23736 B |
|    SerializeIntKeyDict |  1000 | 57,867.2 ns | 594.98 ns | 527.43 ns | 5.6152 |     - |     - |   23736 B |

NET 6.0 Recomputed typeof(string) check 

|                 Method | Count |        Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |------ |------------:|----------:|----------:|-------:|------:|------:|----------:|
| SerializeStringKeyDict |    10 |    868.2 ns |   3.72 ns |   2.90 ns | 0.0801 |     - |     - |     336 B |
|    SerializeIntKeyDict |    10 |    674.7 ns |   7.82 ns |   7.32 ns | 0.0801 |     - |     - |     336 B |
| SerializeStringKeyDict |   100 |  6,688.4 ns | 129.12 ns | 114.46 ns | 0.5035 |     - |     - |    2136 B |
|    SerializeIntKeyDict |   100 |  5,614.3 ns |  98.74 ns | 138.42 ns | 0.5035 |     - |     - |    2136 B |
| SerializeStringKeyDict |  1000 | 68,379.4 ns | 714.07 ns | 596.28 ns | 5.6152 |     - |     - |   23736 B |
|    SerializeIntKeyDict |  1000 | 53,911.5 ns | 671.56 ns | 560.78 ns | 5.6152 |     - |     - |   23736 B |

NET 6.0 Cached typeof(string) check

|                 Method | Count |        Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |------ |------------:|----------:|----------:|-------:|------:|------:|----------:|
| SerializeStringKeyDict |    10 |    858.1 ns |  12.13 ns |  11.35 ns | 0.0801 |     - |     - |     336 B |
|    SerializeIntKeyDict |    10 |    685.8 ns |   3.04 ns |   2.84 ns | 0.0801 |     - |     - |     336 B |
| SerializeStringKeyDict |   100 |  6,764.1 ns |  43.74 ns |  36.53 ns | 0.5035 |     - |     - |    2136 B |
|    SerializeIntKeyDict |   100 |  5,441.2 ns |  78.91 ns |  69.95 ns | 0.5035 |     - |     - |    2136 B |
| SerializeStringKeyDict |  1000 | 69,873.0 ns | 531.20 ns | 496.89 ns | 5.6152 |     - |     - |   23736 B |
|    SerializeIntKeyDict |  1000 | 53,828.0 ns | 267.63 ns | 250.34 ns | 5.6152 |     - |     - |   23736 B |

Reply:
The issue is that if a custom converter for any supported dictionary key type --  string, int32, bool, char, datetime, datetimeoffset, decimal, double, enums, guid, int16, int64, sbyte, single, uint16, uint32, uint64  (exhaustive list) -- the serializer regresses from supporting serializing the type as a dict key using the built-in converters in .NET 5, to throwing `NotSupportedException` in .NET 6.0.

To mitigate this regression, there are a few options:

1. Reverse the regression for only the most popular dictionary key type (string). This PR currently takes this approach, with the intention to support the other types using option 3 in .NET 7. For the other types in 6.0, if custom logic is needed, there are two workarounds:
    - If possible for the scenario, don't register the converter for the types globally using `JsonSerializerOptions.Converters`. Instead, register the converter on the individual primitive properties across the object graph using `[JsonConverter(Type)]`.
    - Register a custom converter for the enclosing dictionary type.
2. Reverse the regression for all the supported types by inlining similar checks as the one for string. This will have an impact on perf, and roots all the built-in converters. This will undo all the source-gen work done to only include converters needed by input object graphs. I don't think we should go down this path.
3. Abandon this PR and instead pursue implementing the "Allow custom converters for dictionary keys" feature in .NET 6.0 - https://github.com/dotnet/runtime/issues/46520. It's a bit late for .NET 6, but it will provide the best mitigation for the regression in a trim-friendly way, though users will still need to write custom code to avoid the regression.

Option 3 is pretty straightforward since we'll just need to remove the `ref WriteStack state` parameter from this [method](https://github.com/dotnet/runtime/blob/cd1b4cff818a700b6ff82fde024c65b036979894/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L596) and make it public. Can pursue this now if you agree.

Reply:
I think option 3 is worth pursuing in .NET 6 -- it provides a workaround for what seems to be a fairly common usage pattern.